File: server/validation.mjs
export function error(message, status = 400) { const e = new Error(message); e.status = status; return e; }
const date = v => typeof v === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(v) && Number.isFinite(Date.parse(v)) && new Date(v).toISOString().slice(0,10) === v;
export function shipmentInput(raw) {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) throw error('Ungültige Eingabe.');
  const out = {};
  for (const key of ['name','device','tracking']) {
    if (typeof raw[key] !== 'string' || raw[key].trim().length > 120) throw error('Bitte die Textfelder prüfen (maximal 120 Zeichen).');
    out[key] = raw[key].trim();
  }
  if (!out.name) throw error('Bitte einen Namen eingeben.');
  if (!['DHL','FedEx'].includes(raw.carrier)) throw error('Bitte DHL oder FedEx auswählen.');
  if (!['Offboarding','Gerätetausch (4 Jahre)','Sonstiges'].includes(raw.reason)) throw error('Bitte einen gültigen Anlass auswählen.');
  if (!date(raw.shipped) || (raw.arrived !== '' && !date(raw.arrived))) throw error('Bitte ein gültiges Datum eingeben.');
  if (raw.arrived && raw.arrived < raw.shipped) throw error('Das Eingangsdatum darf nicht vor dem Versand liegen.');
  if (typeof raw.id !== 'string' || !/^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/.test(raw.id)) throw error('Ungültige Rücksendung.');
  return {...out, carrier:raw.carrier, reason:raw.reason, shipped:raw.shipped, arrived:raw.arrived, id:raw.id};
}

File: server/validation.py
import datetime
import re


class ApiError(Exception):
    def __init__(self, message, status=400):
        super().__init__(message)
        self.status = status


def valid_date(value):
    if not isinstance(value, str) or not re.fullmatch(r'[0-9]{4}-[0-9]{2}-[0-9]{2}', value):
        return False
    try:
        return datetime.date.fromisoformat(value).isoformat() == value
    except ValueError:
        return False


def valid_version(value):
    return type(value) is int and 1 <= value <= 9007199254740991


def shipment_input(raw):
    out = {}
    for key in ('name', 'device', 'tracking'):
        value = raw.get(key)
        if not isinstance(value, str) or len(value.strip().encode('utf-16-le', errors='surrogatepass')) // 2 > 120 or re.search(r'[\ud800-\udfff]', value):
            raise ApiError('Bitte die Textfelder prüfen (maximal 120 Zeichen).')
        out[key] = value.strip()
    if not out['name']:
        raise ApiError('Bitte einen Namen eingeben.')
    if raw.get('carrier') not in ('DHL', 'FedEx'):
        raise ApiError('Bitte DHL oder FedEx auswählen.')
    if raw.get('reason') not in ('Offboarding', 'Gerätetausch (4 Jahre)', 'Sonstiges'):
        raise ApiError('Bitte einen gültigen Anlass auswählen.')
    if not valid_date(raw.get('shipped')) or (raw.get('arrived') != '' and not valid_date(raw.get('arrived'))):
        raise ApiError('Bitte ein gültiges Datum eingeben.')
    if raw['arrived'] and raw['arrived'] < raw['shipped']:
        raise ApiError('Das Eingangsdatum darf nicht vor dem Versand liegen.')
    if not isinstance(raw.get('id'), str) or not re.fullmatch(r'[a-f0-9]{8}(?:-[a-f0-9]{4}){3}-[a-f0-9]{12}', raw['id']):
        raise ApiError('Ungültige Rücksendung.')
    out.update({key: raw[key] for key in ('id', 'carrier', 'reason', 'shipped', 'arrived')})
    return out
